---
title: "R Notebook of plotting cluster level circos plots"
output: html_notebook
---

```{r Libraries, include=FALSE}
library(Seurat)
library(Signac)
library(tidyverse)
library(BSgenome.Mmusculus.UCSC.mm10)
library(RColorBrewer)
library(future)
library(GenomicRanges)
library(EnsDb.Mmusculus.v79)
library(dendextend)
library(DT)
library(ChIPpeakAnno)
library(Repitools)
library(BiocManager)
library(ChIPseeker)
library(circlize)
library(ComplexHeatmap)
library(GOfuncR)
library(fpc)
library(ggsci)
library(qs)
library(TFBSTools)
library(chromVAR)
library(universalmotif)
source("AdditionalFunctions.R")
```

```{r Read E14 merged data}
s.data.e14 <- qread("../scATAC_data/E14_DownstreamReady.071222.RNA.pos.idents.qs", nthreads = 6)
```

```{r UMAP, fig.width=10, fig.height=8}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/All_clusters_UMAP_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
DimPlot(s.data.e14, pt.size = 0.1,  label = TRUE) + coord_fixed() + NoLegend() + ggtitle("")
dev.off()
```

```{r UMAP with brain regions, fig.width=10, fig.height=8}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/All_clusters_UMAP_brain_region_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
DimPlot(s.data.e14, group.by = 'brain_region', pt.size = 0.1) + coord_fixed() + guides(fill=guide_legend(title='MY NEW TITLE')) + ggtitle("")
dev.off()
```

```{r Plot NT type UMAP, fig.width=10, fig.height=8}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/All_clusters_UMAP_NT_type_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
DimPlot(s.data.e14, group.by = 'NT.type', pt.size = 0.1) + ggtitle("") + coord_fixed()
```

```{r Plot clustering reliability, fig.width=10, fig.height=8}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/All_clusters_reliability_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
FeaturePlot(s.data.e14, features = 'silh.score', pt.size = 0.1, cols=c("orange","darkgreen")) + ggtitle("Cluster reliability") + coord_fixed()
```


```{r Calculate NT type counts and proportions}
nt.type.counts <- table(data.frame(Idents(s.data.e14),s.data.e14$NT.type))
nt.type.prop <- apply(nt.type.counts,1,function(x){
  if(sum(x)>0){
    x/sum(x)
    } else {x}
  })
```


```{r Calculate brain region counts and proportions}
brain.region.counts <- table(data.frame(Idents(s.data.e14),s.data.e14$brain_region))
brain.region.prop <- apply(brain.region.counts,1,function(x){
  if(sum(x)>0){
    x/sum(x)
    } else {x}
  })
```


```{r Calculate dendrogram of all clusters}
e14.cluster.averages <- AverageExpression(s.data.e14, assays="peaks")
cell.dist.e14 <- proxy::dist(e14.cluster.averages$peaks, method="cosine", by_rows=FALSE)
hclust_cells_e14 <- hclust(cell.dist.e14, method="ward.D2")
hcd_e14 <- as.dendrogram(hclust_cells_e14)
hcd_e14 <- dendextend::set(hcd_e14,what="branches_lwd", value=8)
```


```{r Plot all cluster dendrogam with Circos, fig.width=24, fig.height=24}
run.time <- format(Sys.time(), "%d-%m-%y_%Hh%Mmin")
pdf(height = 24, width = 24, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Circos_all_clusters_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

circos.clear()
par(mar = rep(4,4))

circos.initialize("leaf.labels", xlim = c(0, length(labels(hcd_e14))))

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.text(1:length(labels(hcd_e14))-0.5, rep(0, length(labels(hcd_e14))), labels(hcd_e14), col = labels_colors(hcd_e14),
               facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex=1.5)
}, bg.border = NA, track.height = 0.15)
max_height = attr(hcd_e14, "height")

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(value=t(brain.region.prop)[as.numeric(labels(hcd_e14)),],pos=1:nrow(brain.region.counts)-0.5, col=pal_d3("category20")(4))
}, track.height =.15, bg.border = NA)

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(value=t(nt.type.prop)[as.numeric(labels(hcd_e14)),],pos=1:nrow(nt.type.counts)-0.5, col=pal_jco()(9)[-8])
}, track.height =.15, bg.border = NA)

circos.track(ylim = c(0, max_height), panel.fun = function(x, y) {
   circos.dendrogram(hcd_e14, max_height = max_height)
}, track.height =.45, bg.border = NA)

lgd_barplot = packLegend(
   Legend(at = colnames(t(nt.type.prop)[as.numeric(labels(hcd_e14)),]), type = "lines",labels_gp = gpar(col = "black", cex = 2.5), legend_gp = gpar(col=pal_jco()(9)[-8], lwd = 8), title_position = "topleft", title_gp = gpar(cex = 2), title = "Neurotransmitter type", size = unit(4, "mm")),
      Legend(at = colnames(t(brain.region.prop)[as.numeric(labels(hcd_e14)),]), type = "lines",labels_gp = gpar(col = "black", cex = 2.5), legend_gp = gpar(col=pal_d3("category20")(4), lwd = 8), title_position = "topleft", title_gp = gpar(cex = 2), title = "Brain region", size = unit(4, "mm"))
     , direction="horizontal")

draw(lgd_barplot, x = unit(2, "cm"), y = unit(2, "cm"), just = c("left", "bottom"))

circos.clear()
dev.off()
```

```{r Calculate some background counts for attributes}
global.nt.counts <- get_attribute_counts(s.data.e14,all.clusters=TRUE, attribute="NT.type")
global.br.counts <- get_attribute_counts(s.data.e14,all.clusters=TRUE, attribute="brain_region")
```

```{r NT type rel enrichment}
nt.res <- lapply(1:10, function(k){
  prunned_dends <- prune_cutree_to_dendlist(hcd_e14,k)
  res.tb.list <- lapply(1:length(prunned_dends),function(d){
    # NT.type
    tmp.nt.counts <- get_attribute_counts(s.data.e14,clusters=labels(prunned_dends[[d]]), attribute="NT.type")
    relevant.global.nt.counts <- global.nt.counts[names(tmp.nt.counts)]
    nt.rel.enr <- (tmp.nt.counts/sum(tmp.nt.counts))/(relevant.global.nt.counts/sum(relevant.global.nt.counts))
    
    # Format results
    res <- c(nt.rel.enr)
    res.tb <- tibble(attribute=names(res),rel.enr=res,prunned_tree=d)
    return(res.tb)
    })
    return(data.table::rbindlist(res.tb.list))
})

```

```{r Brain region relative enrichment}
brain.res <- lapply(1:10, function(k){
  prunned_dends <- prune_cutree_to_dendlist(hcd_e14,k)
  res.tb.list <- lapply(1:length(prunned_dends),function(d){
    
    # Brain region
    tmp.br.counts <- get_attribute_counts(s.data.e14,clusters=labels(prunned_dends[[d]]), attribute="brain_region")
    relevant.global.br.counts <- global.br.counts[names(tmp.br.counts)]
    br.rel.enr <- (tmp.br.counts/sum(tmp.br.counts))/(relevant.global.br.counts/sum(relevant.global.br.counts))
    
    # Format results
    res <- c(br.rel.enr)
    res.tb <- tibble(attribute=names(res),rel.enr=res,prunned_tree=d, clusters=list(as.integer(labels(prunned_dends[[d]]))))
    return(res.tb)
    })
    return(data.table::rbindlist(res.tb.list))
})
```

```{r Construct plot for brain region enrichments}
pl.1.region <- lapply(1:length(brain.res),function(r){
  if (r==1){
      ggplot(brain.res[[r]], aes(fill=attribute, y=rel.enr, x=prunned_tree)) + geom_bar(position="stack", stat="identity") + scale_fill_brewer(palette = "Set3") + theme(axis.text.x = element_blank(), legend.text = element_text(colour="black", size = 6, face = "bold"),legend.key.size = unit(.25, 'cm')) + xlab("") 
  } else {
    new.idents <- construct_ident_vector(tree.branches = brain.res[[r]],old.idents=as.numeric(Idents(s.data.e14)))
    s.data.e14 <- AddMetaData(s.data.e14, new.idents, col.name = "branch")
    d.plot <- DimPlot(s.data.e14, group.by = "branch") + coord_fixed() + ggtitle("")
    
    b.plot <- ggplot(brain.res[[r]], aes(fill=attribute, y=rel.enr, x=prunned_tree)) + geom_bar(position="stack", stat="identity") + scale_fill_brewer(palette = "Set3") + theme(legend.position="none") + xlab("")  + scale_x_continuous(breaks=1:r, labels=as.character(1:r))
    patchwork::wrap_plots(d.plot, b.plot, nrow = 1)
  }
})
```

```{r Construct plot for NT type enrichments}
pl.1.nt.type <- lapply(1:length(nt.res),function(r){
  if (r==1){
      ggplot(nt.res[[r]], aes(fill=attribute, y=rel.enr, x=prunned_tree)) + geom_bar(position="stack", stat="identity") + scale_fill_brewer(palette = "Set3") + theme(legend.text = element_text(colour="black", size = 6, face = "bold"),legend.key.size = unit(.25, 'cm')) + xlab("") + scale_x_continuous(breaks=1:r, labels=as.character(1:r))
  } else {
        new.idents <- construct_ident_vector(tree.branches = brain.res[[r]],old.idents=as.numeric(Idents(s.data.e14)))
    s.data.e14 <- AddMetaData(s.data.e14, new.idents, col.name = "branch")
    d.plot <- DimPlot(s.data.e14, group.by = "branch", pt.size = 0.1) + coord_fixed() + ggtitle("") + ylab("") + xlab("") + theme(legend.text = element_text(size=6), legend.key.size = unit(.025, 'cm'), axis.text.x = element_blank(), axis.text.y = element_blank(), plot.margin =unit(c(.01,.01,.01,.01), "cm"))
    
      b.plot <- ggplot(nt.res[[r]], aes(fill=attribute, y=rel.enr, x=prunned_tree)) + geom_bar(position="stack", stat="identity") + scale_fill_brewer(palette = "Set3") + theme(legend.position="none", plot.margin =unit(c(.01,.01,.01,.01), "cm")) + xlab("")  + scale_x_continuous(breaks=1:r, labels=as.character(1:r))
    patchwork::wrap_plots(b.plot, d.plot, nrow = 1)
  }
})
```

```{r Setting layouts}
layout.left = "
A#########
BB########
CCC#######
DDDD######
EEEEE#####
FFFFFF####
GGGGGGG###
HHHHHHHH##
IIIIIIIII#
JJJJJJJJJJ
"

layout.right = "
#########A
########BB
#######CCC
######DDDD
#####EEEEE
####FFFFFF
###GGGGGGG
##HHHHHHHH
#IIIIIIIII
JJJJJJJJJJ
"
```

```{r Rel.enric. region plots for manuscript}
pdf(file = "../analysis/tree.branching.region.up.to.10.pdf", width = 8, height = 15, onefile = TRUE)
patchwork::wrap_plots(pl.1.region[1:10], ncol = 1, design = layout.right)
dev.off()
```

```{r Rel.enric. NT-type plots for manuscript}
pdf(file = "../analysis/tree.branching.nt.type.up.to.10.pdf", width = 8, height = 15, onefile = TRUE)
patchwork::wrap_plots(pl.1.nt.type[1:10], ncol = 1, design = layout.left)
dev.off()
```


# GABA only

```{r Subset all GABA cells}
reliable.clusters <- as.numeric(tibble(cluster=Idents(s.data.e14), silh=s.data.e14$silh.score) %>% group_by(cluster) %>% summarise(mean.silh=mean(silh)) %>% dplyr::filter(mean.silh > 0.5) %>% pull(cluster) %>% unique())

GABA.e14.data <- subset(s.data.e14, idents=intersect(which(nt.type.prop["GABAergic neurons",]>0.75),reliable.clusters))
```

```{r GABA selector averages}
DefaultAssay(GABA.e14.data) <- "RNA_name"
GABA.e14.data.avg <- AverageExpression(GABA.e14.data)
GABA.fate.selectors <- c("Ascl1","Dlx1","Dlx2","Gata2","Gata3","Tal2","Tal1","Gad1","Gad2")
```

```{r Hierarchical clustering of GABAergic clusters}
cell.dist.gaba <- proxy::dist(GABA.e14.data.avg$peaks, method="cosine", by_rows=FALSE)
hclust_cells.gaba <- hclust(cell.dist.gaba, method="ward.D2")
hcd_gaba <- as.dendrogram(hclust_cells.gaba)
hcd_gaba <- dendextend::set(hcd_gaba, "branches_lwd",6)
```

```{r Setting labels and colors for GABA circos}
cluster.order <- labels(hcd_gaba)
e14.selector.exp <- as.matrix(((t(GABA.e14.data.avg$RNA_name[GABA.fate.selectors,cluster.order]))))
col_fun = colorRamp2(c(min(e14.selector.exp), mean(e14.selector.exp), max(e14.selector.exp)), c("blue", "white", "red"))
```

```{r Draw GABA only circos dendrogram, fig.height=14, fig.width=14}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 24, width = 24, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Circos_GABA_clusters_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

circos.clear()
par(mar = rep(3,4))

circos.par(gap.after = c(10))

circos.heatmap(e14.selector.exp, cluster=FALSE, col = col_fun,  dend.side="none",cell.border = "black", rownames.side = "outside", rownames.cex=3,  track.height = 0.3)

circos.track(track.index = get.current.track.index(), panel.fun = function(x, y) {
    if(CELL_META$sector.numeric.index == 1) { # the last sector
        cn = rev(colnames(e14.selector.exp))
        n = length(cn)
        circos.text(rep(CELL_META$cell.xlim[2], n) + convert_x(1, "mm"), 
            1:n - 0.5, cn, 
            cex = 2.2, adj = c(0, 0.5), facing = "inside")
    }
}, bg.border = NA)


circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(value=t(brain.region.prop)[as.numeric(labels(hcd_gaba)),],pos=1:nrow(brain.region.counts[as.numeric(labels(hcd_gaba)),])-0.5, col=pal_d3("category20")(4))
}, track.height =.08, bg.border = NA)


circos.track(ylim = c(0, max_height), panel.fun = function(x, y) {
   circos.dendrogram(hcd_gaba, max_height = max_height)
}, track.height =.2, bg.border = NA)


lgd_barplot = packLegend(
       Legend(at = colnames(t(brain.region.prop)[as.numeric(labels(hcd_gaba)),]), type = "lines",labels_gp = gpar(col = "black", cex = 2.5), legend_gp = gpar(col=pal_d3("category20")(4), lwd = 8), title_position = "topleft", title_gp = gpar(cex = 2), title = "Brain region", size = unit(4, "mm")),
       Legend(title = "Marker expression", col_fun = col_fun,title_gp = gpar(cex = 2), labels_gp = gpar(fontsize=16))
     , direction="horizontal")

draw(lgd_barplot, x = unit(2, "cm"), y = unit(2, "cm"), just = c("left", "bottom"))

circos.clear()
dev.off()
```

```{r Look at DA features between specific clusters}
# Read results
da_features <- qread("../scATAC_data/da_features_8.qs", nthreads = 6)

da_bp_di_gaba <- lapply(da_features[c(122,18,19,60)],function(da.1){
  lapply(da.1[c(122,18,19,60)], function(da.2){
    if (is_tibble(da.2)){
      filt.da <- da.2 %>% dplyr::filter((avg_log2FC > 0.75 | avg_log2FC < -0.75) & p_val_adj < 0.05) %>% pull(feature)
      
      if (length(filt.da)>0){
        filt.da <- sum(width(StringToGRanges(filt.da)))
      } else {
        filt.da <- 0
      }
    } else {
      filt.da <- 0
    }
  })
})

dalength.matrix <- sapply(da_bp_di_gaba, unlist)/1000
rownames(dalength.matrix) <- as.character(c(122,18,19,60))
colnames(dalength.matrix) <- as.character(c(122,18,19,60))
```

```{r Pairwise chromatin DA count}
da_count <- lapply(da_features[c(122,18,19,60,21)],function(da.1){
  lapply(da.1[c(122,18,19,60,21)], function(da.2){
    if (is_tibble(da.2)){
      filt.da <- da.2 %>% dplyr::filter((avg_log2FC > 0.75 | avg_log2FC < -0.75) & p_val_adj < 0.05) %>% pull(feature)
      if (length(filt.da)>0){
        filt.da <- length(filt.da)
      } else {
        filt.da <- 0
      }
    } else {
      filt.da <- 0
    }
  })
})

dacount.matrix <- sapply(da_count, unlist)
rownames(dacount.matrix) <- as.character(c(122,18,19,60,21))
colnames(dacount.matrix) <- as.character(c(122,18,19,60,21))
```

```{r Draw clustered heatmap of pairwise DA results, fig.width=4, fig.height=4}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 6, width = 6, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/DA_features_heatmap_DI_clusters_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

col_fun = colorRamp2(c(0, median(dalength.matrix),max(dalength.matrix)), c("white","yellow", "red"))

Heatmap(dalength.matrix, col = col_fun, heatmap_legend_param = list(legend_height = unit(6, "cm"),title="kbp",grid_width = unit(1, "cm"),labels_gp = gpar(col = "black", font = 20)), cluster_rows=TRUE, cluster_columns = TRUE)
```

```{r Calculate number of open features per cluster}
binarized.features <- BinarizeCounts(s.data.e14[["peaks"]]@data)

open.per.cluster <- sapply(levels(Idents(s.data.e14)), function(x){
  i <- which(Idents(s.data.e14)==x)
  bin.subset <- binarized.features[,i]
  sum(rowSums(bin.subset)>0)
})
open.per.cluster

open.30.per.cluster <- sapply(levels(Idents(s.data.e14)), function(x){
  i <- which(Idents(s.data.e14)==x)
  bin.subset <- binarized.features[,i]
  sum((rowSums(bin.subset)/ncol(bin.subset))>0.3)
})
open.30.per.cluster


open.50.per.cluster <- sapply(levels(Idents(s.data.e14)), function(x){
  i <- which(Idents(s.data.e14)==x)
  bin.subset <- binarized.features[,i]
  sum((rowSums(bin.subset)/ncol(bin.subset))>0.5)
})
open.50.per.cluster

```

```{r Draw closest gene circoses for cluster pairs,fig.height=15, fig.width=15}
id.1 <- 19
id.2 <- 122

DefaultAssay(s.data.e14) <- "peaks"

dat.1 <- prepare_comp_circos_data(da_features = da_features, id.1=id.1,id.2=id.2, s.data=s.data.e14, avg_log2FC.limit = 0.7)

gene.counts <- dat.1 %>% group_by(label) %>% count()

dat.2 <- dplyr::filter(dat.1, distance <= 50000) %>% distinct(label, .keep_all = TRUE)

dat.2 <- left_join(dat.2, gene.counts)

dat.2$gene_plot_label <- paste(dat.2$label, " (", dat.2$n, ")",sep="")

circos.clear()
circos.initializeWithIdeogram(plotType = c("labels", "axis"),species="mm10")
circos.genomicIdeogram(species="mm10")
circos.genomicLabels(bed=as.data.frame(dat.2), labels.column = 9, side="inside", cex=1, col=dat.2$color, line_col = dat.2$color)

lgd_barplot = packLegend(
   Legend(at = c(id.1,id.2), type = "lines",labels_gp = gpar(col = "black", cex = 1), legend_gp = gpar(col=c("red","blue"), lwd = 4), title_position = "topleft", title_gp = gpar(cex = 1), title = "Cluster", size = unit(2, "mm")))

draw(lgd_barplot, x = unit(2, "cm"), y = unit(2, "cm"), just = c("left", "bottom"))

circos.clear()
```
```{r GeneOntology test for closest genes 122}
clust.122 <- data.frame(genes=dplyr::filter(dat.1, color=="blue") %>%  pull(label),cat=1)

Go_Enrich_Out<- go_enrich(clust.122,organismDb='Mus.musculus')

refined.122 = as_tibble(refine(Go_Enrich_Out, fwer=0.25)) %>% dplyr::filter(signif==TRUE)

create_dt(refined.122)
```

```{r GeneOntology test for closest genes 18}
clust.18 <- data.frame(genes=dplyr::filter(dat.1, color=="blue") %>%  pull(label),cat=1)

Go_Enrich_Out<- go_enrich(clust.18,organismDb='Mus.musculus')

refined.18 = as_tibble(refine(Go_Enrich_Out, fwer=0.25)) %>% dplyr::filter(signif==TRUE)

create_dt(refined.18)
```

# Transcriptome of DI cells

```{r Expression space of Dlx clusters}
Dlx.GABA.e14.data <- subset(s.data.e14, idents=c(122,18,19,60))
DefaultAssay(Dlx.GABA.e14.data) <- "RNA"
Dlx.GABA.e14.data <- FindVariableFeatures(object = Dlx.GABA.e14.data, assay = "RNA", nfeatures = 3000)
Dlx.exp.mat <- as.matrix(FetchData(Dlx.GABA.e14.data, vars = VariableFeatures(Dlx.GABA.e14.data)))
```

```{r Dlx clusters heatmap, fig.height=5, fig.width=20}
genes.to.highlight <- unique(c("Zbtb7c","Ubash3b","Npy", "Isl1", "Scg2", "Sox6", "Six3", "Sst", "Ecel1", "Lhx6", "Dlx2", "Foxg1", "Nkx2-1", "Lhx1", "Pax6", "Prox1", "Nkx2-1", "Foxg1", "Isl1", "Reln", "Prox1", "Six6", "Lhx8", "Arl4d"))
genes.to.highlight.ens <- rownames(Dlx.GABA.e14.data[["RNA"]][[]])[match(genes.to.highlight,Dlx.GABA.e14.data[["RNA"]][[]]$feature_symbol)]
genes.to.i <- match(genes.to.highlight.ens,colnames(Dlx.exp.mat))

col_ha <- HeatmapAnnotation(mark = anno_mark(at = genes.to.i, labels = genes.to.highlight, side="bottom",labels_gp=gpar(fontsize=20)), which = "column")

row_ha <- rowAnnotation(scATAC_cluster = Idents(Dlx.GABA.e14.data), col = list(scATAC_cluster = c("18" = "red", "19" = "green", "60" = "steelblue", "122" = "orange")),annotation_name_gp= gpar(fontsize = 20), annotation_legend_param = list(title = "Cluster", labels_gp=gpar(fontsize=16), title_gp = gpar(fontsize=16)))

# "65" = "grey", "95" = "yellow"

col_fun = colorRamp2(c(0, mean(Dlx.exp.mat),max(Dlx.exp.mat)), c("blue","white", "red"))

run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 24, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Dlx_GABA_transcriptome_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

ht2 <- Heatmap(Dlx.exp.mat, col=col_fun, show_row_names = FALSE, show_column_names = FALSE, right_annotation = row_ha, split=4, clustering_distance_rows = "euclidean", clustering_method_rows = "ward.D2", clustering_distance_columns = "euclidean", clustering_method_columns = "ward.D2", heatmap_legend_param = list(title = "Expression", labels_gp=gpar(fontsize=16), title_gp = gpar(fontsize=16)), bottom_annotation = col_ha, show_column_dend = FALSE, row_title_gp=gpar(fontsize=16))

draw(ht2)
dev.off()
# left_annotation = rowAnnotation(cluster.names = anno_block(gp = gpar(fill =  c("steelblue", "orange","green","red")[c(1,2,3,4)]),labels = c("P3", "MGE", "Ret.Th./LGN", "GABA+NP")[c(1,2,3,4)], labels_gp = gpar(col = "black", fontsize = 10)))
```
```{r DE genes between 4 Dlx clusters}
dlx.cluster.markers <- FindAllMarkers(Dlx.GABA.e14.data, assay = "RNA_name")
dlx.cluster.markers.tb <- as_tibble(dlx.cluster.markers) %>% filter((avg_log2FC > 0.5 | avg_log2FC < -0.5) & p_val_adj <0.05)
create_dt(dlx.cluster.markers.tb)
```

```{r Expression space of R1 clusters}
R1.GABA.e14.data <- subset(s.data.e14, idents=c(74,91,115,53,58,57,68))
DefaultAssay(R1.GABA.e14.data) <- "RNA"
R1.GABA.e14.data <- FindVariableFeatures(object = R1.GABA.e14.data, assay = "RNA", nfeatures = 3000)
R1.exp.mat <- as.matrix(FetchData(R1.GABA.e14.data, vars = VariableFeatures(R1.GABA.e14.data)))
```

```{r R1 clusters heatmap, fig.height=10, fig.width=20}
genes.to.highlight <- unique(c("Zbtb7c","Ubash3b","Npy", "Isl1", "Scg2", "Sox6", "Six3", "Sst", "Ecel1", "Lhx6", "Dlx2", "Foxg1", "Nkx2-1", "Lhx1", "Pax6", "Prox1", "Nkx2-1", "Foxg1", "Isl1", "Reln", "Prox1", "Six6", "Lhx8", "Arl4d"))
genes.to.highlight.ens <- rownames(R1.GABA.e14.data[["RNA"]][[]])[match(genes.to.highlight,R1.GABA.e14.data[["RNA"]][[]]$feature_symbol)]
genes.to.i <- match(genes.to.highlight.ens,colnames(R1.exp.mat))

col_ha <- HeatmapAnnotation(mark = anno_mark(at = genes.to.i, labels = genes.to.highlight, side="bottom"), which = "column")

row_ha <- rowAnnotation(scATAC.cluster = Idents(R1.GABA.e14.data), col = list(scATAC.cluster = c("74" = "red", "115" = "green", "91" = "steelblue", "53" = "orange","58" = "grey", "57"="purple", "68" = "yellow")))

col_fun = colorRamp2(c(0, mean(R1.exp.mat),max(R1.exp.mat)), c("blue","white", "red"))

Heatmap(R1.exp.mat, col=col_fun, show_row_names = FALSE, show_column_names = FALSE, right_annotation = row_ha, split=7, clustering_distance_rows = "euclidean", clustering_method_rows = "ward.D2", clustering_distance_columns = "euclidean", clustering_method_columns = "ward.D2", heatmap_legend_param = list(title = "Expression"), bottom_annotation = col_ha)

#left_annotation = rowAnnotation(cluster.names = anno_block(gp = gpar(fill =  c("steelblue", "orange","green","red","grey","purple","yellow")[c(2,5,7,1,6,4,3)]),labels = c("l R1 GABA neur.", "d R1 GABA precur.", "early/migr.", "d R1 GABA (DTg)", "dl R1 GABA precur.", "d R1 GABA neur. (Foxp2)", "l R1 GABA precur.")[c(2,5,7,1,6,4,3)], labels_gp = gpar(col = "black", fontsize = 6)))

```

```{r DE genes between 7 R1 clusters}
r1.cluster.markers <- FindAllMarkers(R1.GABA.e14.data, assay = "RNA_name")
r1.cluster.markers.tb <- as_tibble(r1.cluster.markers) %>% dplyr::filter((avg_log2FC > 0.5 | avg_log2FC < -0.5) & p_val_adj <0.05)
create_dt(r1.cluster.markers.tb)
```

# VIA part

```{r Subset GABA data for VIA}
GABA.prog.e14.data <- subset(s.data.e14, idents= intersect(c(c(which(nt.type.prop["GABAergic neurons",]>0.75),c(90,21,104))), reliable.clusters))
```

```{r Recalculate LSI and UMAP for DI subset}
DefaultAssay(GABA.prog.e14.data) <- "peaks"
GABA.prog.e14.data <- RunTFIDF(GABA.prog.e14.data)
GABA.prog.e14.data <- FindTopFeatures(GABA.prog.e14.data, min.cutoff = "q0")
GABA.prog.e14.data <- RunSVD(GABA.prog.e14.data, n = 60)

GABA.prog.e14.data <-  RunUMAP(GABA.prog.e14.data,dims = 2:60, reduction = 'lsi')
p1 <- DimPlot(GABA.prog.e14.data, label = TRUE, pt.size=1.2, label.size = 6) + NoLegend() + coord_fixed()
p2 <- DimPlot(GABA.prog.e14.data, label = FALSE, pt.size=1.2, label.size = 6, group.by = "NT.type") + coord_fixed()
p3 <- DimPlot(GABA.prog.e14.data, label = FALSE, pt.size=1.2, label.size = 6, group.by = "brain_region") + coord_fixed()
cowplot::plot_grid(p1,p2,p3)
```

```{r Write out UMAP coordinates for VIA}
VIA.out <- cbind(tibble(barcode=colnames(GABA.prog.e14.data),label = as.character(Idents(GABA.prog.e14.data)), UMAP_1 = GABA.prog.e14.data@reductions$umap@cell.embeddings[,1], UMAP_2 = GABA.prog.e14.data@reductions$umap@cell.embeddings[,2]),as_tibble(GABA.prog.e14.data@reductions$lsi@cell.embeddings))

progenitor.patterning <- c("Sox2","Nkx2-1","Nkx2-2", "Nkx6-1", "Otx2")
GABAergic <- c("Sox21","Ascl1","Gad1","Gad2", "Slc32a1")
GABA.precursor.fate <- c("Helt","Gata2","Gata3","Tal1","Lhx6","Dlx1","Dlx2","Arx")
glutamatergic <- c("Neurog2","Pou4f1","Gls","Slc17a6")
cholinergic <- c("Phox2b","Isl1","Chat","Slc18a3")
dopaminergic <- c("Foxa1","Ddc","Th","Slc18a2")
serotonergic <- c("Pitx3", "Lmx1a","Fev","Tph2","Slc6a4")
adrenergic <- c("En1", "Phox2a","Dbh")

marker_genes <- unique(c(progenitor.patterning,GABAergic, GABA.precursor.fate, glutamatergic, cholinergic, dopaminergic, serotonergic, adrenergic))

DefaultAssay(GABA.prog.e14.data) <- "RNA_name"
GABA.prog.e14.data.markers <- FetchData(GABA.prog.e14.data, vars = marker_genes)

VIA.out <- tibble(cbind(VIA.out, GABA.prog.e14.data.markers))

write.table(VIA.out, file="../analysis/VIA.out.GABA.opt2.csv", sep = ",")
```

# ENCODE comparison

```{r Load ENCODE data for E14}
Combined.brain.E14.cCREs <- readRDS(file="../metadata/Combined.brain.E14.cCREs.Rds")
DefaultAssay(s.data.e14) <- "peaks"
E14.features <- StringToGRanges(rownames(s.data.e14))
```

```{r, fig.height=24, fig.width=24}
ol <- findOverlapsOfPeaks(E14.features, unique(Combined.brain.E14.cCREs), connectedPeaks = "keepAll", minoverlap = 1)

pdf(file="/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/VENN_BRAIN_ENCODE_vs_E14_all.pdf", width = 15, height = 15)
E14_vs_ENCODE.plot <- makeVennDiagram(ol, NameOfPeaks = c("scATAC features","ENCODE cCREs"))
dev.off()
```

```{r Plot annoPie for E14 features not overlapping ENCODE ones}
edb <- EnsDb.Mmusculus.v79
seqlevelsStyle(edb) <- "UCSC"
peakAnno.edb.E14.feat <- annotatePeak(ol$peaklist$E14.features, tssRegion=c(-3000, 3000),TxDb=edb, annoDb="org.Mm.eg.db")

pdf(file="/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/AnnoPie_E14_not_overlapping_ENCODE.pdf", width = 10, height = 8)
plotAnnoPie(peakAnno.edb.E14.feat)
dev.off()
```

```{r Plot annoPie for E14 features overlapping ENCODE ones}
edb <- EnsDb.Mmusculus.v79
seqlevelsStyle(edb) <- "UCSC"
peakAnno.edb.E14.feat.over <- annotatePeak(ol$peaklist$`E14.features///unique.Combined.brain.E14.cCREs.`, tssRegion=c(-3000, 3000),TxDb=edb, annoDb="org.Mm.eg.db")

pdf(file="/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/AnnoPie_E14_overlapping_ENCODE.pdf", width = 10, height = 8)
plotAnnoPie(peakAnno.edb.E14.feat.over)
dev.off()
#vennpie(peakAnno.edb)
```