---
title: R Notebook of studies into defining joint neuronal scATAC feature space for  mouse CNS tissues E12-E15
output:  html_notebook
---

```{r Libraries, include=FALSE}
library(Seurat)
library(Signac)
library(tidyverse)
library(BSgenome.Mmusculus.UCSC.mm10)
library(RColorBrewer)
library(future)
library(GenomicRanges)
library(EnsDb.Mmusculus.v79)
library(org.Mm.eg.db)
library(dendextend)
library(DT)
library(ChIPpeakAnno)
library(Repitools)
library(BiocManager)
library(ChIPseeker)
library(circlize)
library(ComplexHeatmap)
library(genomation)
library(fpc)
library(ggsci)
library(qs)
library(gUtils)
source("AdditionalFunctions.R")
```

```{r Setting multicore/multisession}
options(parallelly.makeNodePSOCK.setup_strategy = "sequential", future.globals.maxSize = 12 * 1024 ^ 3, future.seed=TRUE)
plan("multisession", workers = 6)
```

```{r Path to raw data}
raw.path <- "/Volumes/ExtSSD/ExtraWorkspace/"
run.date <- "180622"
set.seed(2021)
```

# Defining joint feature space

```{r Reading in Seurat object from each separate sample}
# These contain merged technical replicates, read count over 5kbp binned mm10 genome
e12r1.merged <- readRDS("../scATAC_data/E12_R1.merged.clade.phase.170622.Rds")
e14di.merged <- readRDS("../scATAC_data/E14DI1.merged.clade.phase.160622.Rds")
e14mb.merged <- readRDS("../scATAC_data/E14vMB.merged.clade.phase.160622.Rds")
e14vr1.merged <- readRDS("../scATAC_data/E14vR1.merged.clade.phase.170622.Rds")
```

```{r Checking that each have exactly the same amount of features (bins)}
e12r1.merged
e14di.merged
e14mb.merged
e14vr1.merged
```

```{r Creating new merged Seurat object from all of samples}
system.time(neuronal.merged <- merge(x = e12r1.merged, y = c(e14di.merged, e14mb.merged, e14vr1.merged), add.cell.ids = c("_e12r1","_e14di", "_e14vmb","_e14vr1")))
```

```{r Checking the size, and presence of fragment file links of fully merged object}
neuronal.merged
str(Fragments(neuronal.merged))
```

```{r Add gene annotations from Ensembl, include=FALSE}
# extract gene annotations from EnsDb
annotations <- readRDS("../metadata/Ensembel_Mmusculus79_annotation.Rds")

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"

# add the gene information to the object
Annotation(neuronal.merged) <- annotations
```

```{r Calculate nucleosome signal}
neuronal.merged <- NucleosomeSignal(object = neuronal.merged)
```
```{r Plotting Nucleosome signal}
neuronal.merged$nucleosome_group <- ifelse(neuronal.merged$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
FragmentHistogram(object = neuronal.merged, group.by = 'nucleosome_group', region = 'chr1-1-10000000')
```

```{r TSS enrichment}
neuronal.merged <- TSSEnrichment(neuronal.merged, fast = FALSE)
```

```{r Plot biological QC data, fig.width=14}
neuronal.merged$pct_reads_in_peaks <- neuronal.merged$peak_region_fragments / neuronal.merged$passed_filters * 100
neuronal.merged$blacklist_ratio <- neuronal.merged$blacklist_region_fragments / neuronal.merged$peak_region_fragments

VlnPlot(
  object = neuronal.merged,
  features = c('pct_reads_in_peaks', 'peak_region_fragments',
               'TSS.enrichment', 'blacklist_ratio', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5
)

```

```{r Storing merged clade level object as it takes so long to form}
#saveRDS(neuronal.merged,paste("neuronal.merged.clade.phase.",run.date,".Rds",sep=""))
qsave(neuronal.merged, paste("neuronal.merged.clade.phase.",run.date,".qs",sep=""), nthreads = 6)
```

```{r Filter based on biological QC values}
neuronal.merged.filt <- subset(
  x = neuronal.merged,
  subset = peak_region_fragments > 2500 &
    peak_region_fragments < 45000 &
    pct_reads_in_peaks > 55 &
    blacklist_ratio < 0.015 &
    nucleosome_signal < 3 &
    TSS.enrichment > 4.5
)
neuronal.merged.filt
```

```{r Filtering based on number of cells per feature or vice versa}
min.cells.thr <- round(length(unique(Cells(neuronal.merged.filt)))*0.025, digits = 0)
max.cells.thr <- round(length(unique(Cells(neuronal.merged.filt)))*0.975, digits = 0)

min.features.thr <- round(nrow(neuronal.merged.filt)*0.025, digits = 0)
max.features.thr <- round(nrow(neuronal.merged.filt)*0.975, digits = 0)

cell.per.feature <- rowSums(GetAssayData(neuronal.merged.filt[["clade_peaks"]])>0)
feature.per.cell <- colSums(GetAssayData(neuronal.merged.filt[["clade_peaks"]])>0)

feature.filter.index <- which(cell.per.feature >= min.cells.thr & cell.per.feature <= max.cells.thr)
cell.filter.index <- which(feature.per.cell >= min.features.thr & feature.per.cell <= max.features.thr)

features.to.keep <- rownames(neuronal.merged.filt)[feature.filter.index]
cells.to.keep <- Cells(neuronal.merged.filt)[cell.filter.index]

neuronal.merged.filt <- subset(neuronal.merged.filt, cells = cells.to.keep, features = features.to.keep)
neuronal.merged.filt
```


```{r RUN TFIDF for the purpose of clade detection}
neuronal.merged.filt[["clade_peaks_count"]] <- neuronal.merged.filt[["clade_peaks"]]
DefaultAssay(neuronal.merged.filt) <- 'clade_peaks'
neuronal.merged.filt <- BinarizeCounts(neuronal.merged.filt, assay = "clade_peaks")
neuronal.merged.filt <- RunTFIDF(neuronal.merged.filt)
neuronal.merged.filt <- FindTopFeatures(neuronal.merged.filt, min.cutoff = 'q5')
neuronal.merged.filt <- RunSVD(object = neuronal.merged.filt)
```

```{r Save filtered neuronal merged data}
qsave(neuronal.merged.filt, paste("neuronal.merged.clade.phase.filt.",run.date,".qs",sep=""), nthreads = 6)
```

```{r Add replicate genotype and embryo info for batch effect checking}
replicate <- neuronal.merged.filt$replicate

# Which E12R1
e12r1.il <- grepl(names(replicate), pattern = "^_e12r1_.*")
replicate[e12r1.il] <- paste("_e12r1",replicate[e12r1.il], sep="")

# Which E14DI
e14di.il <- grepl(names(replicate), pattern = "^_e14di_.*")
replicate[e14di.il] <- paste("_e14di",replicate[e14di.il], sep="")

# Which E14MB
e14mb.il <- grepl(names(replicate), pattern = "^_e14vmb_.*")
replicate[e14mb.il] <- paste("_e14mb",replicate[e14mb.il], sep="")

# Which E14R1
e14r1.il <- grepl(names(replicate), pattern = "^_e14vr1_.*")
replicate[e14r1.il] <- paste("_e14r1",replicate[e14r1.il], sep="")

genotype <- ifelse(replicate %in% c("_e14di_0","_e14mb_0","_e14r1_0"), "Pax7cre, TdT", "WT")
neuronal.merged.filt <- AddMetaData(neuronal.merged.filt, metadata=paste(replicate,genotype,sep="_"), col.name = "repl.genotype")
neuronal.merged.filt <- AddMetaData(neuronal.merged.filt, metadata=genotype, col.name = "genotype")

embryo <- rep("",length(replicate))
embryo.batch <- rep("",length(replicate))

embryo.batch[replicate %in% c("_e14mb_0", "_e14di_0","_e14r1_0")] <- "6.8.20"
embryo[replicate %in% c("_e14mb_0", "_e14di_0")] <- "6.8.20_E14_embryo1"
embryo[replicate %in% c("_e14r1_0")] <- "6.8.20_E14_embryo2"
embryo[replicate %in% c("_e12r1_0")] <- "6.10.20_E12_embryo1"
embryo[replicate %in% c("_e12r1_1")] <- "6.10.20_E12_embryo2"

embryo.batch[replicate %in% c("_e14mb_1", "_e14di_1","_e14r1_1","_e14mb_2", "_e14di_2","_e14r1_2","_e12r1_0","_e12r1_1")] <- "6.10.20"
embryo[replicate %in% c("_e14mb_1", "_e14di_1","_e14r1_1")] <- "6.10.20_E14_embryo1"
embryo[replicate %in% c("_e14mb_2", "_e14di_2","_e14r1_2")] <- "6.10.20_E14_embryo2"

neuronal.merged.filt <- AddMetaData(neuronal.merged.filt, metadata=embryo.batch, col.name = "embryo.batch")

neuronal.merged.filt <- AddMetaData(neuronal.merged.filt, metadata=embryo, col.name = "embryo")
```

```{r Correlation between depth and reduced dimension components}
DepthCor(neuronal.merged.filt)
```

```{r Take approximate right singular vectors, matrix multiplication with singular values and transpose}
SVD.d <- neuronal.merged.filt@reductions$lsi@misc$d
SVD.u <- neuronal.merged.filt@reductions$lsi@misc$u
num.dimensions <- ncol(neuronal.merged.filt@reductions$lsi)
d_diagtsne = matrix(0, nrow=num.dimensions, ncol=num.dimensions)
diag(d_diagtsne) = SVD.d
# Assuming that first principle component is associated with read depth as shown above.
d_diagtsne[1,1] = 0
# Take approximate right singular vectors, matrix multiplication with singular values and transpose
SVDtsne_vd = t(d_diagtsne %*% t(SVD.u))
cell.dist <- proxy::dist(SVDtsne_vd, method="cosine")
hclust_cells <- hclust(cell.dist, method="ward.D2")
hcd <- as.dendrogram(hclust_cells)
```

```{r Calculate cluster statistics}
# Calculation km.stats with cluster.stats (fpc package) at various cut levels at HPC cluster and returning here
km.stats.10 <- qread("../analysis/km.stats.res.10.200622.qs")
km.stats.9.5 <- qread("../analysis/km.stats.res.9.5.200622.qs")
km.stats.9 <- qread("../analysis/km.stats.res.9.200622.qs")
km.stats.8.5 <- qread("../analysis/km.stats.res.8.5.200622.qs")
km.stats.8 <- qread("../analysis/km.stats.res.8.200622.qs")
km.stats.7.5 <- qread("../analysis/km.stats.res.7.5.200622.qs")
km.stats.7 <- qread("../analysis/km.stats.res.7.200622.qs")
km.stats.6.5 <- qread("../analysis/km.stats.res.6.5.200622.qs")
km.stats.6 <- qread("../analysis/km.stats.res.6.200622.qs")
km.stats.5.5 <- qread("../analysis/km.stats.res.5.5.200622.qs")
km.stats.5 <- qread("../analysis/km.stats.res.5.200622.qs")
km.stats.4.5 <- qread("../analysis/km.stats.res.4.5.200622.qs")
km.stats.4 <- qread("../analysis/km.stats.res.4.200622.qs")

km.stats.list <- list(km.stats.10=km.stats.10, km.stats.9.5=km.stats.9.5, km.stats.9=km.stats.9, km.stats.8.5=km.stats.8.5, km.stats.8=km.stats.8, km.stats.7.5=km.stats.7.5, km.stats.7=km.stats.7, km.stats.6.5=km.stats.6.5, km.stats.6=km.stats.6, km.stats.5.5=km.stats.5.5, km.stats.5=km.stats.5, km.stats.4.5=km.stats.4.5,km.stats.4=km.stats.4)
```

```{r Plot clustering statistics}
avg.silhouette <- sapply(km.stats.list, function(x){x$avg.silwidth})
dunn2 <- sapply(km.stats.list, function(x){x$dunn2})

clusterint.stats <- tibble(clus.param=as.numeric(str_remove(names(dunn2), pattern = "km.stats.")), dunn2=dunn2, avg.silhouette=avg.silhouette)

run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 3, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Clade_level_cluster_parameters_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

ggplot(clusterint.stats, aes(x=clus.param, group=1)) + geom_line(aes(y=dunn2, color="Dunn2")) + theme_minimal() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_y_continuous("Dunn2", sec.axis = sec_axis(~ (. / 2.5), name = "Avg. Silhouette")) + geom_line(aes(y=avg.silhouette*2.5, color="Avg. Silhouette")) + xlab("Clade cutting height") + scale_x_continuous(breaks = seq(4, 10, by = 0.5))  + theme(legend.title=element_blank(), text = element_text(size=16))
dev.off()
```

```{r Cut the tree for purpose of visualization, fig.width=14}
par(mai=c(2,.5,.25,.25),cex=.75)
h =5.5
hcd.cut <- cut(hcd, h = h)
hcd.upper<-hcd.cut$upper
hcd.upper.clusters<-dendextend::cutree(hcd.upper,h=h)
hcd.upper<-dendextend::reindex_dend(hcd.upper)
# Find out why this needs to be done, what is not reset in the object after cut but is reset after conversion back and worth
hcd.upper<-as.dendrogram(as.hclust(hcd.upper))
n.term.nodes <- sapply(hcd.cut$lower,dendextend::count_terminal_nodes)
leaf.labels <- paste("Clade #",hcd.upper.clusters,": ",n.term.nodes," cells",sep="")
hcd.upper<-dendextend::set_labels(hcd.upper,leaf.labels)
#color_branches(hcd.upper,clusters=hcd.upper.clusters,col=color.vector(hcd.upper.clusters)) 
hcd.upper %>% dendextend::set("branches_lwd",5) %>% dendextend::set("labels_cex",1.2) %>% plot(center=T)
abline(h = h, lty = 2, col="red", lwd=2)
```

```{r Cut the tree to find main clades and write those out for summit identification}
cells_tree_cut = cutree(hclust_cells, h=h)
lsi_cells = dplyr::tibble(barcode = Cells(neuronal.merged.filt), cells_tree_cut = cells_tree_cut)
```

```{r See how samples are distributed in clades}
sample.clade <- table(str_extract(string = lsi_cells$barcode, pattern = "_.*__") %>% str_replace_all(pattern = "_*", replacement = ""), lsi_cells$cells_tree_cut)

rownames(sample.clade)[which(rownames(sample.clade)=="e14vmb")] <- "e14mb"
rownames(sample.clade)[which(rownames(sample.clade)=="e14vr1")] <- "e14r1"
sample.clade

sample.clade.prop <- apply(sample.clade,2,function(x){x/sum(x)})

repl.genotype <- table(neuronal.merged.filt$repl.genotype,lsi_cells$cells_tree_cut)
repl.genotype.prop <- apply(repl.genotype,2,function(x){x/sum(x)})

genotype.clade <- table(neuronal.merged.filt$genotype,lsi_cells$cells_tree_cut)
genotype.clade.prop <- apply(genotype.clade,2,function(x){x/sum(x)})

embryo.clade <- table(neuronal.merged.filt$embryo,lsi_cells$cells_tree_cut)
embryo.clade.prop <- apply(embryo.clade,2,function(x){x/sum(x)})

embryo.batch.clade <- table(neuronal.merged.filt$embryo.batch,lsi_cells$cells_tree_cut)
embryo.batch.clade.prop <- apply(embryo.batch.clade,2,function(x){x/sum(x)})
```

```{r Draw circos plot with brain regions stacked barplot, fig.height=24, fig.width=24}
pdf(file="/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Clade_circos_repl_genotype.pdf", width = 24, height = 24)

hcd.upper.circ  <- hcd.upper
par(mar = rep(4,4))
circos.initialize("leaf.labels", xlim = c(0, ncol(sample.clade)))

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.text(1:ncol(sample.clade)-0.5, rep(0, ncol(sample.clade)), labels(hcd.upper.circ), col = dendextend::labels_colors(hcd.upper.circ),
               facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5), cex=1.2)
}, bg.border = NA, track.height = 0.15)
max_height = attr(hcd.upper.circ, "height")

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(t(sample.clade.prop),1:ncol(sample.clade)-0.5, col=ggsci::pal_jco()(4))
}, track.height =.10, bg.border = NA)

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(t(genotype.clade.prop),1:ncol(genotype.clade)-0.5, col=ggsci::pal_npg()(nrow(genotype.clade)))
}, track.height =.10, bg.border = NA)

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(t(embryo.batch.clade.prop),1:ncol(embryo.batch.clade)-0.5, col=ggsci::pal_nejm()(nrow(embryo.batch.clade)))
}, track.height =.10, bg.border = NA)

circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
   circos.barplot(t(embryo.clade.prop),1:ncol(embryo.clade)-0.5, col=ggsci::pal_igv()(nrow(embryo.clade)))
}, track.height =.10, bg.border = NA)

circos.track(ylim = c(0, max_height), panel.fun = function(x, y) {
   circos.dendrogram(hcd.upper.circ, max_height = max_height)
}, track.height =.30, bg.border = NA)

lgd_barplot = packLegend(
  Legend(at = rownames(sample.clade.prop), type = "points",labels_gp = gpar(col = "black", cex = 2), legend_gp = gpar(col=ggsci::pal_jco()(4), lwd = 2), title_position = "topleft", title_gp = gpar(cex = 2), title = "Brain region", size = unit(6, "mm")),
    Legend(at = rownames(genotype.clade.prop), type = "points",labels_gp = gpar(col = "black", cex = 2), legend_gp = gpar(col=ggsci::pal_npg()(nrow(genotype.clade)), lwd = 2), title_position = "topleft", title_gp = gpar(cex = 2), title = "Genotype", size = unit(6, "mm")),
      Legend(at = rownames(embryo.clade.prop), type = "points",labels_gp = gpar(col = "black", cex = 2), legend_gp = gpar(col=ggsci::pal_igv()(nrow(embryo.clade)), lwd = 2), title_position = "topleft", title_gp = gpar(cex = 2), title = "Embryo", size = unit(6, "mm")),
      Legend(at = rownames(embryo.batch.clade.prop), type = "points",labels_gp = gpar(col = "black", cex = 2), legend_gp = gpar(col=ggsci::pal_nejm()(nrow(embryo.batch.clade)), lwd = 2), title_position = "topleft", title_gp = gpar(cex = 2), title = "Batch date", size = unit(6, "mm"))
    )

draw(lgd_barplot, x = unit(2, "cm"), y = unit(2, "cm"), just = c("left", "bottom"))
dev.off()
```

# Macs2 and peak detection

```{r Run MACS for clade level peak detection, include=FALSE}
neuronal.merged.filt <- AddMetaData(neuronal.merged.filt, lsi_cells$cells_tree_cut, "clade")
clade.peaks <- CallPeaks(neuronal.merged.filt, macs2.path = "/Users/kilpinen/opt/anaconda3/envs/r411_221021/bin/macs2", effective.genome.size = 1.87e9, group.by = "clade", combine.peaks = TRUE, name = paste(sample.name,"_per_clade_peaks_merged", sep=""))
```

```{r Just sanity check of peaks called over the genome}
clade.peaks
```

```{r Store clade peaks for further use}
saveRDS(clade.peaks,paste("../analysis/joint.clade.peaks.",run.date,".Rds",sep=""))
```

# Looking for correspondence between joint feature space and separate ones

```{r Loading separately generated feature spaces}
clade.peaks.di <- readRDS("../analysis/E14DI_clade.peaks.Rds")
clade.peaks.e12r1 <- readRDS("../analysis/E12R1_clade.peaks.191021.Rds")
clade.peaks.e14vmb1 <- readRDS("../analysis/E14MB_clade.peaks.Rds")
clade.peaks.e14vr1 <- readRDS("../analysis/E14_R1.021121.cladepeaks.Rds")
```

```{r}
clade.peaks
```

There are 482 157 features in jointly defined feature space

```{r Stats of features, fig.width=12}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 12, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Feature_length_histogram",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
ggplot(tibble(feat.widths=(ranges(clade.peaks)@width)),aes(x=feat.widths)) + geom_histogram(bins=50) + theme_minimal() + xlab("Feature length in bp") + ylab("Count") + theme(text = element_text(size=26)) + scale_x_continuous(breaks = seq(0, 6000, by = 500))
stat_bin(binwidth=250, geom='text', color='white', aes(label=..count..), position=position_stack(vjust = 0.5))
dev.off()
```

```{r Peak locations circos, fig.height=8, fig.width=8}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 12, width = 12, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Feature_density_circos_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))
circos.initializeWithIdeogram(plotType = c("labels", "axis"),species="mm10",ideogram.height = convert_height(5, "mm"),axis.labels.cex = 0.8*par("cex"), labels.cex = 1.2*par("cex"))
circos.genomicIdeogram(species="mm10")
circos.genomicDensity(Repitools::annoGR2DF(clade.peaks), col="#EFC000FF")
dev.off()
```

```{r Feature distances to nearest TSS}
# Let's use Chipseeker package

edb <- EnsDb.Mmusculus.v79
seqlevelsStyle(edb) <- "UCSC"
promoters <- getPromoters(TxDb=edb, upstream=3000, downstream=3000, by="gene")

tagMatrix <- getTagMatrix(clade.peaks, windows=promoters)

plotAvgProf(tagMatrix, xlim=c(-3000, 3000), xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency", conf = 0.95, resample = 1000)
```

```{r}
average.feature.loc.gene <- plotPeakProf2(peak = clade.peaks, upstream = rel(0.25), downstream = rel(0.25),
              conf = 0.95, by = "gene", type = "body", nbin = 250,
              TxDb = edb, ignore_strand = F, free_y=FALSE)
average.feature.loc.gene + theme(text=element_text(size=26))
```

```{r Plot pie charts of feature annotation categories}
peakAnno.edb <- annotatePeak(clade.peaks, tssRegion=c(-3000, 3000),TxDb=edb, annoDb="org.Mm.eg.db")

plotAnnoPie(peakAnno.edb)
vennpie(peakAnno.edb)
```

```{r Distribution of features relative to TSS, fig.height=2.5, fig.width=8}
plotDistToTSS(peakAnno.edb, title="Distribution of features relative to TSS")
```

```{r Plotting random granges}
random.peaks <- shuffle(clade.peaks, TxDb=BSgenome.Mmusculus.UCSC.mm10)

peakAnno.rand.edb <- annotatePeak(random.peaks, tssRegion=c(-3000, 3000),TxDb=edb, annoDb="org.Mm.eg.db")

plotAnnoPie(peakAnno.rand.edb)
```


```{r Comparison to E14.5 brain cCREs from ENCODE}
ENCODE.Fore_cCREs <- readGeneric("../metadata/Fore_ENCFF237YNH_ENCFF349REI_ENCFF279EVI.7group.bed", remove.unusual = TRUE, header=FALSE, keep.all.metadata = TRUE)
ENCODE.Fore_cCREs <- ENCODE.Fore_cCREs[which(!elementMetadata(ENCODE.Fore_cCREs)[,ncol(elementMetadata(ENCODE.Fore_cCREs))-1] %in% c("Low-DNase", "Unclassified"))]

ENCODE.MB_cCREs <- readGeneric("../metadata/MB_ENCFF539CBL_ENCFF829ZIT_ENCFF104QAC.7group.bed", remove.unusual = TRUE, header=FALSE, keep.all.metadata = TRUE)
ENCODE.MB_cCREs <- ENCODE.MB_cCREs[which(!elementMetadata(ENCODE.MB_cCREs)[,ncol(elementMetadata(ENCODE.MB_cCREs))-1] %in% c("Low-DNase", "Unclassified"))]

ENCODE.Hind_cCREs <- readGeneric("../metadata/Hind_ENCFF274NPS_ENCFF475FUT_ENCFF081VBW.7group.bed", remove.unusual = TRUE, header=FALSE, keep.all.metadata = TRUE)
ENCODE.Hind_cCREs <- ENCODE.Hind_cCREs[which(!elementMetadata(ENCODE.Hind_cCREs)[,ncol(elementMetadata(ENCODE.Hind_cCREs))-1] %in% c("Low-DNase", "Unclassified"))]

ENCODE.E12.Hind_cCREs <- readGeneric("../metadata/E12_hind_ENCFF344PSC_ENCFF333XRC.7group.bed", remove.unusual = TRUE, header=FALSE, keep.all.metadata = TRUE)
ENCODE.E12.Hind_cCREs <- ENCODE.E12.Hind_cCREs[which(!elementMetadata(ENCODE.E12.Hind_cCREs)[,ncol(elementMetadata(ENCODE.E12.Hind_cCREs))-1] %in% c("Low-DNase", "Unclassified"))]

Combined.brain.cCREs <- do.call("c", GRangesList(ENCODE.Fore_cCREs, ENCODE.MB_cCREs, ENCODE.Hind_cCREs,ENCODE.E12.Hind_cCREs))
saveRDS(Combined.brain.cCREs, file="../metadata/Combined.brain.cCREs.Rds")

Combined.brain.E14.cCREs <- do.call("c", GRangesList(ENCODE.Fore_cCREs, ENCODE.MB_cCREs, ENCODE.Hind_cCREs))
saveRDS(Combined.brain.E14.cCREs, file="../metadata/Combined.brain.E14.cCREs.Rds")
```

```{r}
ggplot(tibble(feat.widths=ranges(unique(Combined.brain.cCREs))@width),aes(x=feat.widths)) + geom_histogram(bins=50) + theme_minimal() + xlab("cCREs length in bp") + ylab("Count") + theme(text = element_text(size=20)) + scale_x_continuous(breaks = seq(100, 350, by = 50)) + ggtitle("ENCODE mouse brain cCREs")
```


```{r, fig.width=10, fig.height=6}
ggplot(tibble(feat.widths=ranges(clade.peaks)@width),aes(x=feat.widths)) + geom_histogram(bins=50) + theme_minimal() + xlab("Feature length in bp") + ylab("Count") + theme(text = element_text(size=20)) + scale_x_continuous(breaks = seq(0, 5000, by = 500))  + ggtitle("Features in scATAC space")
```

```{r, fig.height=24, fig.width=24}
pdf(file="VENN_BRAIN_ENCODE_vs_all.pdf", width = 15, height = 15)
makeVennDiagram(Peaks=list(clade.peaks,unique(Combined.brain.cCREs)), NameOfPeaks = c("scATAC features","ENCODE cCREs"), minoverlap = 1)
dev.off()
```



Vs. for example 287 587 features in separately defined E14Di feature space.

```{r Hist. of E14di separate feature space feature overlaps with the joint one}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/E14DI_features_in_joint_feature_space_histogram_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

hits <- findOverlaps(clade.peaks, clade.peaks.di)
overlaps <- pintersect(clade.peaks[queryHits(hits)], clade.peaks.di[subjectHits(hits)])
percentOverlap <- width(overlaps) / width(clade.peaks.di[subjectHits(hits)])
ggplot(tibble(overlap=percentOverlap),aes(x=overlap)) + geom_histogram(colour="black", fill=pal_jco()(6)[6]) + theme_minimal() + xlab("Proportion (length) of feature-to-feature overlap") + ggtitle("") + theme(text = element_text(size=14)) + ylab("n(features")
dev.off()
#hits <- hits[percentOverlap > 0.8]
```

```{r Hist. of joint feature space feature overlaps with the E14DI separate one}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 8, width = 8, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/Joint_features_in_E14DI_feature_space_histogram_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

hits <- findOverlaps(clade.peaks.di, clade.peaks)
overlaps <- pintersect(clade.peaks.di[queryHits(hits)], clade.peaks[subjectHits(hits)])
percentOverlap <- width(overlaps) / width(clade.peaks[subjectHits(hits)])
ggplot(tibble(overlap=percentOverlap),aes(x=overlap)) + geom_histogram(colour="black", fill=pal_jco()(6)[6]) + theme_minimal() + xlab("Proportion (length) of feature-to-feature overlap") + ggtitle("Hist. of joint feature space feature overlaps\nwith the E14DI separate one") + theme(text = element_text(size=14)) + ylab("n(features")
dev.off()
```

```{r VENN diagram Joint vs E14DI vs E14vMB, fig.height=6, fig.width=6}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 6, width = 6, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/VENN_diagram_Joint_vs_E14DI_vs_E14vMB_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

joint.E14DI.E14vMB.plot <- makeVennDiagram(Peaks=list(clade.peaks,clade.peaks.di, clade.peaks.e14vmb1), NameOfPeaks = c("Joint space","E14DI", "E14vMB"), minoverlap = 100)
dev.off()
```

```{r VENN diagram of E12R1 vs E14R1 vs joint space, fig.height=6, fig.width=6}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 6, width = 6, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/VENN_diagram_E12R1_vs_E14R1_vs_Joint_space_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

makeVennDiagram(Peaks=list(clade.peaks,clade.peaks.e14vr1, clade.peaks.e12r1), NameOfPeaks = c("Joint space","E14vR1","E12R1"), minoverlap = 100)
dev.off()
```

```{r VENN diagram with all comparisons, fig.height=6, fig.width=6}
run.time <- format(Sys.time(), "%d-%m-%y")
pdf(height = 15, width = 15, file=paste("/Users/kilpinen/OneDrive - University of Helsinki/Neuronal Feature Space/Figures/plots for figures/VENN_diagram_all_comparisons_",run.time,".pdf",sep=""))
pushViewport(viewport(gp = gpar(fontfamily = "Helvetica")))

makeVennDiagram(Peaks=list(clade.peaks,clade.peaks.di, clade.peaks.e14vmb1, clade.peaks.e12r1,clade.peaks.e14vr1), NameOfPeaks = c("Joint space","E14DI", "E14vMB","E12R1","E14vR1"), minoverlap = 100)
dev.off()
```

```{r Env info}
sessionInfo()
```

